staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -489,25 +508,28 @@ func (qs *queueSet) shuffleShardLocked(hashValue uint64, descr1, descr2 interfac
 	offset := qs.enqueues % handSize
 	qs.enqueues++
 	bestQueueIdx := -1
-	bestQueueSeatsSum := int(math.MaxInt32)
+	minQueueSeatSeconds := MaxSeatSeconds
 	for i := 0; i < handSize; i++ {
 		queueIdx := hand[(offset+i)%handSize]
 		queue := qs.queues[queueIdx]
-		waitingSeats := queue.requests.SeatsSum()
-		// TODO: Consider taking into account `additional latency` of requests
-		// in addition to their seats.
-		// Ideally, this should be based on projected completion time in the
+		queueSum := queue.requests.QueueSum()
+
+		// the best queue is chosen based on projected completion time in the
 		// virtual world of the youngest request in the queue.
-		thisSeatsSum := waitingSeats + queue.seatsInUse
-		klog.V(7).Infof("QS(%s): For request %#+v %#+v considering queue %d of %d seats waiting and %d executing, nextDispatchR=%v", qs.qCfg.Name, descr1, descr2, queueIdx, waitingSeats, queue.seatsInUse, queue.nextDispatchR)
-		if thisSeatsSum < bestQueueSeatsSum {
-			bestQueueIdx, bestQueueSeatsSum = queueIdx, thisSeatsSum
+		thisQueueSeatsSum := queueSum.MaxSeatsSum + queue.seatsInUse
+
+		// this is the total amount of work in seat-seconds either in progress or
+		// waiting in this queue, we will select the queue with the minimum.
+		thisQueueSeatSeconds := SeatsTimesDuration(float64(thisQueueSeatsSum), qs.estimatedServiceDuration) + queueSum.AdditionalSeatSecondsSum
```

Comment:
> I think that the score used for shuffle-sharding should be the sum of the work in the queue's executing and waiting requests.

I'm not sure I agree. For requests that are already executing, they may potentially be almost finished already, in which case your algorithm will look this queue much worse than it can be in practice.
We should try to project the expected remaining amount of work instead (which we can do based on the R() meter and the value.


That said - if we assume we want to have what you want (which I'm temporarily fine as this is the logic we have currently anyway) - then it should be:

SeatsTimesDuration(queueSum.InitialSeatsSum(), qs.estimatedServiceDuration) + queueSum.AdditionaSeatsSecondsSum + queue.seatsInUse * something (which we actually can't say what it is, because we're not tracing in which phase a request is - so assuming qs.estimatedServiceDuration is somewhat reasonable guess).

Reply:
okay, fixed.

Reply:
I would personally suggest going with the first option from the above list and merging this PR.

I agree it's not ideal - but given it's roughly what we had so far, I don't perceive it as a regression, even though it's not exactly what we want.

This would unblock this PR and let us make further progress, and we can discuss in the meantime how we can improve that further.

@MikeSpreitzer - WDYT?